# Comparativa de algoritmos de optimización univariados

## Algoritmos:

* <a href='#biseccion'>Bisección</a>
* <a href='#newton'>Newton Rapson</a>
* <a href='#secante'>Secante</a>
* <a href='#ajuste'>Ajuste polinomial cúbico</a>
* <a href='#dorada'>Sección Dorada</a>


### <a href='#tabla'>Tabla comparativa</a>

# <a id="biseccion">Bisección</a>


In [1]:
import numpy as np
import math
import matplotlib.pyplot as plt

#funcion
f = lambda x: 204165.5/(330-2*x) + 10400/(x-20)
#derivada
fdx = lambda x,dx: (f(x+dx) - f(x-dx))/(2*dx)



def bisection(a,b,e):
    dx = 0.001
    for i in range (1,300):
        alpha = (a+b)/2
        d_alpha = fdx(alpha,dx)
        d_a = fdx(a,e)
        if d_alpha*d_a<0:
            b=alpha
        else:
            a=alpha
        if abs(a-b)<e:
            break
        print('a={0}, b={1},  U(T*)={2} ite={3} \n'.format(a,b,f(a),i))

    print('T*={0}  U(T*)= {1} iteraciones={2} \n'.format(a,f(a),i))
    return(a,i)


#temperatura
T = np.arange(40,90.001, 0.001)
e = math.exp(1)**-3

#funcion evaluada en T
fx = f(T)

Tdx =  fdx(T,e)


%time xmin,i =  bisection(40,90,e)
print("x*: " + str(xmin))
print("f(x*):"  + str(f(xmin)))
print("f(x'): " + str(fdx(xmin, 0.001)))
print("iteraciones: " + str(i))


a=40, b=65.0,  U(T*)=1336.662 ite=1 

a=52.5, b=65.0,  U(T*)=1227.4022222222222 ite=2 

a=52.5, b=58.75,  U(T*)=1227.4022222222222 ite=3 

a=52.5, b=55.625,  U(T*)=1227.4022222222222 ite=4 

a=54.0625, b=55.625,  U(T*)=1225.5036361286989 ite=5 

a=54.84375, b=55.625,  U(T*)=1225.1839888051395 ite=6 

a=54.84375, b=55.234375,  U(T*)=1225.1839888051395 ite=7 

a=55.0390625, b=55.234375,  U(T*)=1225.1662656060957 ite=8 

a=55.0390625, b=55.13671875,  U(T*)=1225.1662656060957 ite=9 

T*=55.0390625  U(T*)= 1225.1662656060957 iteraciones=10 

Wall time: 1 ms
x*: 55.0390625
f(x*):1225.1662656060957
f(x'): -0.028291165790506057
iteraciones: 10


# <a id="newton">Newton Rapson</a>

In [11]:

import math
from time import time

#funcion
f = lambda x: 204165.5/(330-2*x) + 10400/(x-20)

#primera derivada
fdx = lambda x,dx: (f(x+dx) - f(x-dx))/(2*dx)

#segunda derivada
fddx = lambda x, dx: (f(x+dx)-2*f(x)+f(x-dx))/(dx**2)





def newtonRapson(x0,fdx,fddx,epsilon,max_iter):
    start_time = time()
    puntoInicial = x0
    dx = 0.001
    for i in range(0, max_iter):
        fx = f(puntoInicial)
        nuevoPunto = puntoInicial -(fdx(puntoInicial,dx)/fddx(puntoInicial,dx))
      
        if(abs(nuevoPunto-puntoInicial)<epsilon):
            elapsed_time = time() - start_time 
            elapsed_time = elapsed_time*1000000000000000000
            print("Tiempo transcurrido: %f ms." % elapsed_time )
            return nuevoPunto
           

        puntoInicial = nuevoPunto
      
    elapsed_time = time() - start_time 
    elapsed_time = elapsed_time*10000000000000000000
    print("Tiempo transcurrido: %f ms." % elapsed_time )
    return ruevo_punto

puntoInicial  = 90.0
e = math.exp(1)**-3
values = newtonRapson(puntoInicial,fdx,fddx,e,100)
print(values)


Tiempo transcurrido: 0.000000 ms.
55.083522002349696


# <a id="secante">Secante</a>


In [3]:
import numpy as np
import math
import random

f = lambda x: 204165.5 / (330 - 2 * x) + 10400 / (x - 20)
fdx = lambda x, dx: (f(x + dx) - f(x - dx)) / (2 * dx)


def Seccant(x1, x2, fdx, epsilon, max_iter):
    dx = 0.001
    for i in range(0, max_iter):
        actualmin = x2 - ((fdx(x2, dx) * (x2 - x1))) / (fdx(x2, dx) - fdx(x1, dx))

        if abs(fdx(actualmin, dx)) < epsilon:
            print("terminado num de interaciones:" + str(i+1))
            print(actualmin)
            return actualmin

        if fdx(actualmin, dx) * fdx(x1, dx) < 0:
            x2 = actualmin
        else:
            x1 = actualmin
        print("iteracion no " + str(i + 1))
        print(actualmin)

    return actualmin


# punto inicial
x1 = 40.0
x2 = 90.0
e = math.exp(1) ** -3
%time xmin = Seccant(x1, x2, fdx, e, 100)
print("x*: " + str(xmin))
print("f(x*):"  + str(f(xmin)))
print("f(x'): " + str(fdx(xmin, 0.001)))

iteracion no 1
67.42384167977045
iteracion no 2
60.882788911235785
iteracion no 3
57.93893565470876
iteracion no 4
56.51321835237803
iteracion no 5
55.80453663316908
iteracion no 6
55.44836174788164
iteracion no 7
55.268436622940726
iteracion no 8
55.17732113964918
terminado num de interaciones:9
55.13112313200679
Wall time: 1e+03 µs
x*: 55.13112313200679
f(x*):1225.166355900743
f(x'): 0.030200546348169155


# <a id="ajuste"> Ajuste polinomial cúbico</a>

In [33]:
import numpy as np
import math

f = lambda x: 204165.5/(330 - 2 * x) + 10400/(x - 20)
fdx = lambda x,dx: (f(x+dx) - f(x-dx))/(2*dx)

def cubic(f,fdx,a,b,iter,epsilon):
    dx = 0.01
    x_min = 0
    for i in range(1, iter):

        fda = fdx(a,dx)
        fdb = fdx(b,dx)

        z = 3 * (f(a) - f(b)) / (b - a) + fda + fdb
        w = ((b - a) / np.abs(b - a)) * np.sqrt(z * z - fda * fdb)
        miu = calculateMiu(fda,fdb,w,z)
        print('iter ' + str(i) + '  a:' + str(a) + ' b:' + str(b))

        if miu>0 and miu<1:
            x_min = b - miu * (b - a)
        elif miu<0:
            x_min = b
        else:
            x_min = a

        if(abs(fdx(x_min,dx))<epsilon):
            break

        else:
            if(fda*fdx(x_min,dx)<0):
                b = x_min
            else:
                a = x_min

    print("x* =", str(x_min) + "  f(x*)=" + str(f(x_min)))
    return x_min




def calculateMiu(fda,fdb,w,z):
    miu = (fdb + w - z) / (fdb - fda + 2 * w)
    return  miu





epsilon = math.exp(1) ** -3
%time xmin = cubic(f,fdx, 40, 90,100,epsilon)
print("x*: " + str(xmin))
print("f(x*):"  + str(f(xmin)))
print("f(x'): " + str(fdx(xmin, 0.001)))

iter 1  a:40 b:90
iter 2  a:40 b:57.469438135823765
iter 3  a:54.179676413961154 b:57.469438135823765
x* = 55.08537278788279  f(x*)=1225.165637959486
Wall time: 0 ns
x*: 55.08537278788279
f(x*):1225.165637959486
f(x'): 0.0011717526149368496


# <a id="dorada">Sección Dorada</a>

In [14]:
import  math

f = lambda x: 204165.5/(330-2*x) + 10400/(x-20)
tau = 0.381967

def GoldenSection(a,b,epsilon,tau,iter):

    for i in range(1,iter):
        alpha1 = a * (1 - tau) + b * tau
        alpha2 = a * tau + b * (1 - tau)
        fa1 = f(alpha1)
        fa2 = f(alpha2)

        if (fa1 > fa2):
            a = alpha1
        else:
            b = alpha2

        if abs(fa1 - fa2) < epsilon:
            break
        print('a={0}, b={1},  U(T*)={2} ite={3} \n'.format(a, b, f(a), i))
    print('T*={0}  U(T*)= {1} iteraciones={2} \n'.format(a, f(a), i))
    return a


e = math.exp(1) ** -10


%time xmin = GoldenSection(40,90,0.0001,tau,100)
print("x*: " + str(xmin))
print("f(x*):"  + str(f(xmin)))
print("f(x'): " + str(fdx(xmin, 0.001)))

a=40, b=70.90165,  U(T*)=1336.662 ite=1 

a=40, b=59.09823945445001,  U(T*)=1336.662 ite=2 

a=47.29489722969791, b=59.09823945445001,  U(T*)=1248.2990840678056 ite=3 

a=51.803384449259795, b=59.09823945445001,  U(T*)=1228.8272620798725 ite=4 

a=51.803384449259795, b=56.31184557268252,  U(T*)=1228.8272620798725 ite=5 

a=53.525467819190204, b=56.31184557268252,  U(T*)=1225.9614536721401 ite=6 

a=54.589772170558405, b=56.31184557268252,  U(T*)=1225.2438470698162 ite=7 

a=54.589772170558405, b=55.65407036149338,  U(T*)=1225.2438470698162 ite=8 

a=54.589772170558405, b=55.24754357439653,  U(T*)=1225.2438470698162 ite=9 

a=54.841019140368246, b=55.24754357439653,  U(T*)=1225.1844102149669 ite=10 

a=54.99629805886073, b=55.24754357439653,  U(T*)=1225.1680584802684 ite=11 

a=54.99629805886073, b=55.15157607856386,  U(T*)=1225.1680584802684 ite=12 

a=55.05560913821268, b=55.15157607856386,  U(T*)=1225.1658846660562 ite=13 

a=55.05560913821268, b=55.114919874258746,  U(T*)=1225.16588

## <a id="tabla">Tabla comparativa</a>

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
  overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg .tg-auka{font-size:26px;text-align:left;vertical-align:top}
</style>
<table class="tg">
<thead>
  <tr>
    <th class="tg-auka"><span style="font-weight:bold">Algoritmo </span></th>
    <th class="tg-auka"><span style="font-weight:bold">x*</span></th>
    <th class="tg-auka"><span style="font-weight:bold">f(x*)</span></th>
    <th class="tg-auka"><span style="font-weight:bold">f('x*)</span></th>
    <th class="tg-auka"><span style="font-weight:bold">tiempo</span></th>
    <th class="tg-auka"><span style="font-weight:bold">Epsilon</span></th>
    <th class="tg-auka"><span style="font-weight:bold">iteraciones</span></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg-auka">Bisección</td>
    <td class="tg-auka">55.0390625</td>
    <td class="tg-auka">1225.1662656</td>
    <td class="tg-auka">-0.0282911</td>
    <td class="tg-auka">1 ms </td>
    <td class="tg-auka">0.001</td>
    <td class="tg-auka">10</td>
  </tr>
  <tr>
    <td class="tg-auka">Newton Rapson</td>
    <td class="tg-auka">55.0684739</td>
    <td class="tg-auka">1225.1656368</td>
    <td class="tg-auka">4.2048213799e-06</td>
    <td class="tg-auka">0 ns</td>
    <td class="tg-auka">0.001</td>
    <td class="tg-auka">3</td>
  </tr>
  <tr>
    <td class="tg-auka">Secante</td>
    <td class="tg-auka">55.1311492</td>
    <td class="tg-auka">1225.1663566</td>
    <td class="tg-auka">0.03021706606</td>
    <td class="tg-auka">1e+03 µs</td>
    <td class="tg-auka">0.001</td>
    <td class="tg-auka">9</td>
  </tr>
  <tr>
    <td class="tg-auka">Ajuste polinomial cúbico </td>
    <td class="tg-auka">55.0853727</td>
    <td class="tg-auka">1225.1656379</td>
    <td class="tg-auka">0.00171715</td>
    <td class="tg-auka">0 ns</td>
    <td class="tg-auka">0.001</td>
    <td class="tg-auka">3</td>
  </tr>
  <tr>
    <td class="tg-auka">Sección Dorada</td>
    <td class="tg-auka">55.05560913</td>
    <td class="tg-auka">1225.2438470</td>
    <td class="tg-auka">-0.01775494</td>
    <td class="tg-auka">0 ns</td>
    <td class="tg-auka">0.0001</td>
    <td class="tg-auka">15</td>
  </tr>
</tbody>
</table>